### ISSUE TITLE
Wrong definition of the matching window size causes ValueError: not enough values to unpack (expected 4, got 0).

### ISSUE THREAD

**Issue by danschef** [SMByC/Coregistration-Qgis-processing#4]
The global co-registration fails for some input images with an error which I cannot reproduce when using the AROSICS Python API instead of the QGIS plugin:

```python
QGIS version: 3.18.3-Zürich
Qt version: 5.12.9
GDAL version: 3.3.3
GEOS version: 3.10.0-CAPI-1.16.0
PROJ version: Rel. 8.1.1, September 1st, 2021
PDAL version: 2.3.0 (git-version: 47a710)
Processing algorithm…
Algorithm 'Automated global Co-Registration' starting…
Input parameters:
{ 'ALIGN_GRIDS' : True, 'IMG_REF' : '/test/ref.tif', 'INPUT' : '/test/tgt.tif', 'MATCHING_WINDOW' : None, 'MATCH_GSD' : True, 'MAX_SHIFT' : 5, 'OUTPUT' : 'TEMPORARY_OUTPUT', 'RESAMPLING' : 0 }

Image to image Co-Registration:

Processing file: /home/gfz-fe/scheffler/temp/coreg_bug_james_lawrence_2019/Bournemouth_ee49f7e5-228f-4377-8a93-0a4a46f4f426-inv.tif

Perform automatic subpixel co-registration with Arosics...

(To check the complete log of the process, open the Python Console)...

Traceback (most recent call last):
File "/home/gfz-fe/.local/share/QGIS/QGIS3/profiles/default/python/plugins/Coregistration/automated_global_coregistration_algorithm.py", line 272, in processAlgorithm
CR = COREG(img_ref, img_tgt, path_out=output_file, align_grids=align_grids, match_gsd=match_gsd,
File "/home/gfz-fe/mambaforge/envs/coreg/lib/python3.9/site-packages/arosics/CoReg.py", line 433, in __init__
self._get_clip_window_properties() # sets self.matchBox, self.otherBox and much more
File "/home/gfz-fe/mambaforge/envs/coreg/lib/python3.9/site-packages/arosics/CoReg.py", line 885, in _get_clip_window_properties
overlapPoly_within_matchWin = matchBox.mapPoly.intersection(self.overlap_poly)
File "/home/gfz-fe/mambaforge/envs/coreg/lib/python3.9/site-packages/py_tools_ds/geo/vector/geometry.py", line 92, in mapPoly
return shapelyImPoly_to_shapelyMapPoly(imPoly, self.gt)
File "/home/gfz-fe/mambaforge/envs/coreg/lib/python3.9/site-packages/py_tools_ds/geo/vector/conversion.py", line 53, in shapelyImPoly_to_shapelyMapPoly
xmin, ymin, xmax, ymax = shapelyBox.bounds
ValueError: not enough values to unpack (expected 4, got 0)
```

My guess is that it comes from the way how you define the matching window in case the user does not provide one. [Here](https://github.com/SMByC/Coregistration-Qgis-processing/blob/3fab841e21c1d31b4037de17098c785fbfd9d4bc/automated_global_coregistration_algorithm.py#L254), you set the matching window size to xGSD x yGSD. This is wrong because you are inserting map units there and the matching window size is defined in pixel units. Apart from that, the matching window size is automatically determined in AROSICS, so I suggest to not touch it as long as the user really provides a custom matching window.

**Comment by danschef** [comment]
Also please note that the matching window size refers to the size of the moving window which AROSICS uses to compute the shift at a certain position, NOT to the entire area in which tie points are computed (in case of global co-registration, there is only one matching window). So, the default matching window size is not equal to the image overlap but is set to 256x256 pixels. This should be adapted in the description of the 'custom matching window' parameter. However, having this parameter is nice because it makes it handy to adjust the actual matching window position. But it should really be used wit caution because users might misunderstand it and mix it up with the overlap area. Users might tend to provide much too large matching windows which then negatively affects the co-registration result. For this reason, AROSICS only offers the possibility to customize the window position and size in pixels.

**Comment by XavierCLL** [comment]
My original idea of the `Custom machining window` in the global algorithm is to let the user draw on canvas the overlap area as an option:
![image](https://user-images.githubusercontent.com/11801453/142710750-146eec92-d095-4559-ab7d-1729f3508820.png)

So I use the X and Y boundaries of that rectangle drawn by the user to define the `wp` and the `ws` in Arosics, is that wrong? or am I misunderstanding the  `wp` and the `ws` parameters?

**Comment by danschef** [comment]
In case the user defines a custom matching window:
From the technical point of view, your approach works. However, note that AROSICS computes the shifts between two input images via phase correlation in the frequency domain. The image content of the matching window is transformed to the frequency domain and then used for phase correlation. For this, it is not reasonable to have fourier image sizes of more than 512x512 pixels. Therefore, AROSICS does not compute the shift on the entire image overlap (which,e.g., might be like 8000x8000 for a Landsat image) but only on a small subset of the input image (which I call matching window). Giving the user full freedom of how to define the matching window in the map canvas might mean that the user does not take care on a reasonable matching window size (256x256 pixels by default) and, e.g. define one with 5000x5000 pixels. This is computationally way more expensive AND will lead to an inaccurate computation of the shift. That is why AROSICS only lets the user define the center position of the matching window and gives some advise on its size but not neccessarily requires to set one. So, a more resonable approach in the QGIS plugin would be (1) to let the user click on the map to get the center position of the matching window and then possibly add another (optional) parameter for the size or (2) at least raise a warning for matching window sizes of, lets say 512x512 pixels. However, the idea of drawing the window on the canvas is cool but reeaally has to used with care to get reasonable results.

In case the user does not define a matching window:
[Here](https://github.com/SMByC/Coregistration-Qgis-processing/blob/3fab841e21c1d31b4037de17098c785fbfd9d4bc/automated_global_coregistration_algorithm.py#L254) your current code grabs the pixel size in map units from the gdal GeoTransform tuple and defines the matching window with x_res x y_res. This is wrong and causes the above error because the matching window is not defined in map units but with a pixel dimension. Currently, your version would define a matching window of 30x30 pixels for a Landsat image (ground resolution 30 meters) and override the default of 256x256 pixels. A 30 pixel matching window is usually too small to get resonable results. However, in case the input image comes with geographical coordinates, the x_gsd/y_gsd is way below 1, thats why AROSICS crashes with the above error (I admit, that such small matching window sizes should be rejected by AROSICS, I will improve that).

**Comment by XavierCLL** [comment]
Thanks @danschef for your detailed explanation on this issue!

I've finally opted to choose the first option, to let the user click on the map to get the center position of the matching window and the parameter for the size using 256 as the default value. And of course, I fixed the problem that you mention at the end when the user lets it empty.

If you have the opportunity to test it would be great, thanks!
